### PR TITLE
GPII-3136: Updating GCP project README with CouchDB data restoration guide

### PR DESCRIPTION
**Pull request by natarajaya**

### CODE REVIEW

File: gcp/README.md [hunk 1]
```diff
@@ -77,3 +77,43 @@ This happens when trying to enable an API that is already enabled. This shouldn'
...
 rake xk"[gcloud services disable container.googleapis.com]"
 ```
+
+### Restoring CouchDB data
+
+We are considering number of probable failure scenarios for our GCP infrastructure:
+
+1. **Data corruption on a single CouchDB replica**
+
+In this scenario we rely on CouchDB ability to recover from loss of one or more replicas (our current production CouchDB settings allow us to lose up to 2 random nodes and still keep data integrity). The best course of action as follows:
```
Comment (amatas) [inline gcp/README.md]: I'd say that is allow to lose up to (n_replicas - 1) random nodes.

File: gcp/README.md [hunk 2]
```diff
@@ -77,3 +77,43 @@ This happens when trying to enable an API that is already enabled. This shouldn'
...
+* Get current number of CouchDB stateful set replicas with `kubectl --namespace gpii get statefulset couchdb-couchdb -o jsonpath="{.status.replicas}"`.
+* Scale CouchDB stateful set to 0 replicas with `kubectl --namespace gpii scale statefulset couchdb-couchdb --replicas=0`.
+* This will cause K8s to terminate all CouchDB pods, all PDs that were mounted into them will be released.
+* Open Google Cloud console, go to "Compute Engine" -> "Disks".
+* Now, for every PD you collected:
+   * Remember PD's name, type, size and zone.
+   * Pick proper snapshot.
+   * Delete PD.
+   * Create new PD from snapshot with the same name, type, size and zone.
+* Scale CouchDB stateful set back to number of replicas it used to have before.
```
Comment (amatas) [inline gcp/README.md]: I'd add also the kubctl command for this.

File: gcp/README.md [hunk 3]
```diff
@@ -77,3 +77,43 @@ This happens when trying to enable an API that is already enabled. This shouldn'
...
+* Database is now restored to the state at the time of target snapshot.
```
Comment (amatas) [inline gcp/README.md]: I'd also propose a method to check that the database is working properly. Perhaps mapping ports using kubectl and then making curl requests.

File: gcp/README.md [hunk 4]
```diff
@@ -77,3 +77,43 @@ This happens when trying to enable an API that is already enabled. This shouldn'
...
+* Delete associated PVC and then affected pod. For our example case:
+   * `kubectl --namespace gpii delete pvc database-storage-couchdb-couchdb-1`
+   * `kubectl --namespace gpii delete pod couchdb-couchdb-1`
+* After target pod is terminated, Persistent Disk that was mounted into it thru corresponding PVC will be destroyed as well.
+* When new pod is created to replace deleted one, corresponding PVC will be created as well, and, thru it, new PV object for new GCE PD.
+* Please note that you'll have to patch newly created PV, so `k8s-snapshots` can continue to make its snapshots, for our example case:
+   * Get volume name from `kubectl --namespace gpii get pvc database-storage-couchdb-couchdb-1`.
+   * Patch PV with `kubectl patch pv PV_NAME_FROM_PREVIOUS_STEP -p '{"metadata": {"annotations": {"backup.kubernetes.io/deltas": "ENV_DELTAS"}}}'`
+   * To get proper ENV_DELTAS value, check CouchDB Terragrunt module for the environment you're working with (i.e. `live/prd/k8s/gpii/couchdb/terraform.tfvars` for production environment, with deltas `PT5M PT60M PT24H P7D P52W`).
+* CouchDB cluster will replicate data to recreated node automatically.
```
Comment (amatas) [inline gcp/README.md]: I'd add a check to see that the cluster has the nº of the replicas that is expected, and also a check of each replica to see that the cluster is retrieving the same data from each replica. For example, a kubctl command that map a port for each replica, and the curl check should retrieve the same results from each replica.

File: gcp/README.md [hunk 5]
```diff
@@ -77,3 +77,43 @@ This happens when trying to enable an API that is already enabled. This shouldn'
...
+* Make sure that you figured affected CouchDB pod properly
+* There is a PVC object, associated with affected CouchDB pod. Let's say affected pod is `couchdb-couchdb-1`, then corresponding PVC is `database-storage-couchdb-couchdb-1`, located in the same namespace.
+* Delete associated PVC and then affected pod. For our example case:
+   * `kubectl --namespace gpii delete pvc database-storage-couchdb-couchdb-1`
+   * `kubectl --namespace gpii delete pod couchdb-couchdb-1`
+* After target pod is terminated, Persistent Disk that was mounted into it thru corresponding PVC will be destroyed as well.
+* When new pod is created to replace deleted one, corresponding PVC will be created as well, and, thru it, new PV object for new GCE PD.
+* Please note that you'll have to patch newly created PV, so `k8s-snapshots` can continue to make its snapshots, for our example case:
+   * Get volume name from `kubectl --namespace gpii get pvc database-storage-couchdb-couchdb-1`.
+   * Patch PV with `kubectl patch pv PV_NAME_FROM_PREVIOUS_STEP -p '{"metadata": {"annotations": {"backup.kubernetes.io/deltas": "ENV_DELTAS"}}}'`
```
Comment (mrtyler) [inline gcp/README.md]: Hmm, this is kind of rough. Can we instead run `rake` again (or maybe `rake deploy_module[couchdb]`) and let the terraform code handle this for us?

File: gcp/README.md (hunk 1 above)
Comment (natarajaya) [reply gcp/README.md]: Here is [n value](http://docs.couchdb.org/en/2.1.2/cluster/theory.html) from current CouchDB settings for GCP:
```
/project # kubectl --namespace gpii exec couchdb-couchdb-0 -c couchdb -- cat /opt/couchdb/etc/default.ini | grep -E "n=[0-9]"
n=3```
Which means that we can lose exactly 2 random nodes.

File: gcp/README.md (hunk 2 above)
Comment (natarajaya) [reply gcp/README.md]: Done

File: gcp/README.md (hunk 3 above)
Comment (natarajaya) [reply gcp/README.md]: Done

File: gcp/README.md (hunk 4 above)
Comment (natarajaya) [reply gcp/README.md]: I can only think of something like:
```# rake xk["kubectl exec --namespace gpii -it couchdb-couchdb-0 -c couchdb -- curl -s http://\$TF_VAR_couchdb_admin_username:\$TF_VAR_couchdb_admin_password@127.0.0.1:5984/gpii/"]
docker-compose run --rm --service-ports xk kubectl exec --namespace gpii -it couchdb-couchdb-0 -c couchdb -- curl -s http://$TF_VAR_couchdb_admin_username:$TF_VAR_couchdb_admin_password@127.0.0.1:5984/gpii/
{"db_name":"gpii","update_seq":"209-g1AAAALLeJzLYWBg4MhgTmFwSM4vTc5ISYLRujDaQA9NRC-9IDNTr7gsWS85p7S4JLVILyc_OTEnB2gSUyJDkvz___-zEqWoZ2aSApBMsgcbK4rXWEOSjHUAGRsPNlaWiq5NABlbDzZWmnquzWMBkgwNQApo8vysRDnquRhi9AKI0fsJhTEZRh-AGH2fBq5-ADEaGNaSWQA35vsp","sizes":{"file":268208,"external":150004,"active":118527},"purge_seq":0,"other":{"data_size":150004},"doc_del_count":0,"doc_count":209,"disk_size":268208,"disk_format_version":6,"data_size":118527,"compact_running":false,"cluster":{"q":8,"n":2,"w":2,"r":2},"instance_start_time":"0"}
```
to check DB status on a node. If you have better ideas, please feel free to add them into README and commit into this branch!

File: gcp/README.md (hunk 5 above)
Comment (mrtyler) [reply gcp/README.md]: Cool. Looks like you can replace the top-level bullet at line 96 and all its children with "Run `rake deploy_module[couchdb]`".

Comment (natarajaya) [reply gcp/README.md]: Done

**Pull request merged**